src/libNetwork/P2PComm.cpp

```diff
@@ -360,193 +288,297 @@ bool P2PComm::SendMessageSocketCore(const Peer& peer,
+                    else
+                    {
+                        LOG_GENERAL(WARNING, "Incorrect message hash.");
+                        return;
+                    }
+                }
             }
-            else
+
+            if (found)
             {
-                LOG_GENERAL(WARNING, "Incorrect message hash.");
+                // We already sent and/or received this message before -> discard
+                LOG_GENERAL(INFO, "Discarding duplicate broadcast message.");
                 return;
             }
-        }
-    }
 
-    cli_sock_closer.reset(); // close socket now so it can be reused
-
-    if (found)
-    {
-        // We already sent and/or received this message before -> discard
-        LOG_GENERAL(INFO, "Discarding duplicate broadcast message");
-        return;
-    }
+            unsigned char msg_type = 0xFF;
+            unsigned char ins_type = 0xFF;
+            if (messageLength - HASH_LEN > MessageOffset::INST)
+            {
+                msg_type = message.at(HDR_LEN + HASH_LEN + MessageOffset::TYPE);
+                ins_type = message.at(HDR_LEN + HASH_LEN + MessageOffset::INST);
+            }
 
-    unsigned char msg_type = 0xFF;
-    unsigned char ins_type = 0xFF;
-    if (message.size() > MessageOffset::INST)
-    {
-        msg_type = message.at(MessageOffset::TYPE);
-        ins_type = message.at(MessageOffset::INST);
-    }
+            vector<Peer> broadcast_list
+                = m_broadcast_list_retriever(msg_type, ins_type, *from);
 
-    vector<Peer> broadcast_list
-        = m_broadcast_list_retriever(msg_type, ins_type, from);
+            if (broadcast_list.size() > 0)
+            {
+                P2PComm::GetInstance().RebroadcastMessage(broadcast_list,
+                                                          message, msg_hash);
+            }
 
-    if (broadcast_list.size() > 0)
-    {
-        P2PComm::GetInstance().SendBroadcastMessageCore(broadcast_list, message,
-                                                        msg_hash);
-    }
+            P2PComm::GetInstance().ClearBroadcastHashAsync(msg_hash);
 
-    // Used to be done in SendBroadcastMessageCore, but it would never be called by lookup nodes
-    P2PComm::GetInstance().ClearBroadcastHashAsync(msg_hash);
+            LOG_STATE("[BROAD]["
+                      << std::setw(15) << std::left
+                      << P2PComm::GetInstance().m_selfPeer << "]["
+                      << DataConversion::Uint8VecToHexStr(msg_hash).substr(0, 6)
+                      << "] RECV");
 
-    LOG_STATE(
-        "[BROAD][" << std::setw(15) << std::left
-                   << P2PComm::GetInstance().m_selfPeer << "]["
-                   << DataConversion::Uint8VecToHexStr(msg_hash).substr(0, 6)
-                   << "] RECV");
+            // Move the shared_ptr message to raw pointer type
+            pair<vector<unsigned char>, Peer>* raw_message
+                = new pair<vector<unsigned char>, Peer>(
```

Comment:
Is it possible to use make_pair?

Reply:
Can't use make_pair here since what we want to construct is a heap object, not a local.